### ISSUE TITLE
Consider alternative to 0-1-2 severities

### ISSUE THREAD

**Issue by davidtheclark** [stylelint/stylelint#382]
Somebody whose opinion I trust recently pointed out to me how unintuitive the ESLint 0-1-2 pattern is for specifying severity, and how that obscurity can annoy people who need to modify the codebase but are not super familiar with the linter. I've heard it before from others, too (e.g. https://twitter.com/Hagenburger/status/619128858177077249). 

I thought about it a little bit and I have an alternative that we could consider, better than just replacing 0-1-2 with keywords "ignore"-"warn"-"error". Consider:

- Passing `false` would turn off the rule (same as `0` currently), e.g. `"block-no-empty": false`.
- By default, rules would have an "error"-level severity. I think this is probably what the majority of users will use for the majority of rules. So just passing a setting means turning on the rule with "error"-level severity (and `true` for `-no-*` rules). E.g. `"block-no-empty": true, "declaration-semicolon-space-before": "always"`.
- To downgrade the rule to "warn"-level severity, all rules could accept a new property in their secondary options object. Maybe `warnOnly`, or something like that. E.g. `"block-no-empty": [true, { warnOnly: true }], "at-rule-empty-line-before": ["always-multi-line", { ignore: ["after-comment"], warnOnly: true }]`.

It's not really a big deal, I think, because if the 0-1-2 convention is prominently documented people really have no reason to complain. But I thought I'd throw out this idea, in case anybody liked it.

**Comment by jeddy3** [comment]
> But I thought I'd throw out this idea, in case anybody liked it.

I've kind of gotten use to the whole 0-1-2 pattern and so I'm probably not the best person to comment on this. Lets flag this up for discussion though, in case anyone else has thoughts on it.

**Comment by davidtheclark** [comment]
Adding the v3 milestone; because I think we either implement some change for v3 or we leave as is indefinitely.

**Comment by jeddy3** [comment]
Sounds like a good idea.

**Comment by MoOx** [comment]
What about null, false, true? Just an idea :) Not very intuitive either.
false, "warn", "err" ?
I have to admit that error by default seems a good idea indeed.

**Comment by jeddy3** [comment]
Last of the `3.0.0` milestone issues... 

Thoughts?

**Comment by davidtheclark** [comment]
I like the proposal above, where error is default, "warn" is set as a secondary option, false turns the rule off. Don't want to piss anybody off, though, and haven't heard any other feelings about it.

**Comment by davidtheclark** [comment]
> Are we anticipating having to add this option to all the rules, or can this be handled externally from the rules themselves?

I think we could make this change outside the rule. Currently as you say the rules do not know or care about their severity level. I would have to develop a proof-of-concept to be sure, but I think we could do this without having to do something tedious for all rules.

@TrySound, as the newest org member, what do you think?

Also, I know that @ai and @ben-eb do a lot of thinking about APIs and the PostCSS ecosystem --- do either of you have opinions about this that you'd like to share?

**Comment by davidtheclark** [comment]
Also cc/ @keithamus and @ntwb, who have shared opinions in the past.

**Comment by ai** [comment]
Hm. In one case `false, "warn", "err"` is much closer to ESLint.

In other case, we really should promote `error` level instead of `warning`. Longer API is always a good way for it.

**Comment by davidtheclark** [comment]
Proposal is:

- Error by default
- `false` turns the rule off altogether
- If you want warning level without turning off the rule, you set that as a secondary option, `{ warn: true }`, or `{ warnOnly: true }`.

**Comment by ben-eb** [comment]
Agreed, integer options are confusing. Elegant, but confusing. I like `{warn: true}`.

**Comment by keithamus** [comment]
Personally, eslint parity is important to me. I can totally understand and respect offering more readably alternatives though.

So whats wrong with having both? `"block-no-empty": false` or `"block-no-empty": 0` could both be valid, same as `"block-no-empty": "warn"`/ `"block-no-empty": 1` and `"block-no-empty": "error"`/ `"block-no-empty": 2`.

**Comment by davidtheclark** [comment]
@ben-eb: "Elegant, but confusing" seems contrary to me :)

@keithamus: We could handle both numbers and strings if we didn't change rule config format. But what I'm proposing is that we do away with that initial severity marker in favor defaulting to error. (If we *don't* switch formats and default to error, maybe we still should add those string aliases.)

Instead of `color-hex-case: [2, "upper"]` we'd have `color-hex-case: "upper"`. And instead of `color-hex-case: [1, "upper"]` we'd have `color-hex-case: ["upper", { warn: true }]`.

As I see it, the advantages of changing are:
- No ambiguous number system. This may be the biggest one. As @jeddy3 said: "I can also imagine it being much clearer than the 0, 1, 2 approach if a user is unfamiliar with eslint." I started this thread because my brother first encountered ESLint and was really put off by the cryptic numbering thing, instead of having a self-explanatory config.
- No need for including a whole lots of `2`s in your configuration.
- Error by default.

The only disadvantages I can think of are:
- Defaults to warning instead of forcing user to make a choice (though I'm not convinced by this one, since I'm not super interested in the warning severity).
- ESLint is the big dog and they use numbers.

**Comment by ben-eb** [comment]
Yeah; more like confusing until you understand it. But not so readable, I think. :smile:

**Comment by TrySound** [comment]
We can make `eslintLike` option to enable it. It would be cool to have partial back compatibility. The are many users already use stylelint. And rewrite big config just for new version is not quite good.

And I like `true`, value, `false` notation :)

**Comment by keithamus** [comment]
@davidtheclark with a config like like `color-hex-case: ["upper", { warn: true }]` how do you specify additional rule options? Do they go in the second object, e.g. `color-hex-case: ["upper", { warn: true, exceptions: ['#fff'] }]` (contrived example I know). If that's the case then you're effectively making `warn` and `error` reserved keywords in the options object. Or do you supply options as a third arg? `color-hex-case: ["upper", { warn: true }, { exceptions: ["#fff'] }]`? I think making severity more readable comes at the cost of making the rule options more complex.


The other thing you could do, is flip the whole thing on its head, and have `rules.warnings`, `rules.errors`, so your config might look like:


```js
{
  rules: {
    warnings: {
      "color-hex-case": ["upper", { exceptions: ["#fff"] }],
    }
    errors: {
      "color-hex-length": 3,
    }
  }
}
```

**Comment by keithamus** [comment]
To expand a bit - most of the current stylelint rules only take one argument as an option, but look at a rule like [eslint's id-match](http://eslint.org/docs/rules/id-length.html) which takes several optional arguments. I see no reason why stylelint wouldn't/shouldn't feature complex rules like this - but to do so you'll need to devise a format that can scale to that complexity.

**Comment by davidtheclark** [comment]
Our configuration convention that has worked out really well so far is that each rule requires a primary option, in whatever format it needs, and may take a secondary options object.

```
"some-rule": [severity, primary, secondary]
```

That has held up well and I anticipate no reason to change it. "warn" would go in the same secondary options object as rule-unique options.

> If that's the case then you're effectively making warn and error reserved keywords in the options object.

Just "warn", but yeah that would be the plan. 

> The other thing you could do, is flip the whole thing on its head, and have rules.warnings, rules.errors

That's an interesting thought.

**Comment by davidtheclark** [comment]
@jeddy3: I feel like letting you make the call on this one. I have explained my reasoning & plan and I believe you understand it. Others have offered their thoughts. You could be a good final judge on what we do for v3. What do you think?

**Comment by jeddy3** [comment]
> I feel like letting you make the call on this one. I have explained my reasoning & plan and I believe you understand it. Others have offered their thoughts.

Crikey. That's a tough shout as everyone has raised lots of valid points, advantages and disadvantages to both approaches.

Although an interesting idea, I think we can dismiss `rules.warnings` and  `rules.errors` as I can see it getting a little clunky when extending a config and wanting to turn off a rule i.e. do I need to put the rule within either the `error` or `warning` keys before turning it off?

I think the key advantages to your proposal are that it is _intuitive_ and _errors by default_. Both of which, I feel, would greatly help new users get started more quickly.

Everyone makes valid points about the advantages of the existing approach, from @keithamus's parity with eslint to @ben-eb's confusing to begin with but eloquent once you do. 

But I think the key suggestion and point comes from @TrySound:

>  It would be cool to have partial back compatibility. There are many users who already use stylelint. And rewrite big config just for new version is not quite good.

All these points makes me think that, if it is both technically feasibility and not a monster effort, we adopt the proposal as the default config style (i.e. we use it in the documentation examples and I migrate the likes of `config-suitcss` and the example config over to this style) to assist first-time users, but we allow people to continue using the eslint style behind a flag in the config. 

I'm not sure what that flag is though. Whether @TrySound suggestion of `eslintLike` or something that focuses on the fact that you must explicitly define the severity first e.g. `explicitSeverities: true`. The latter option also captures one of the key reasons behind why we allow two config styles.

This will allow owners of existing configs to simply add this flag to their configs for compatibility with `3.0.0`, while allowing us to put in place a new default config style for new users.

If you agree that is a reasonable solution, then I guess the next step is to develop a proof-of-concept. Both to see if we can work with the `warn` secondary option externally from the rules, and to see whether using a flag in the config is viable.

---

FYI, I'm off walking tomorrow for 10 days (hurray!). So, I won't be able to assistance in this unless we tackle the final stages after the 30th of Nov. As always though, feel free to merge and release in my absence :)

**Comment by davidtheclark** [comment]
Thanks for the thoughtful response, @jeddy3. I think we could do this without much headache. I'll make a PR.

Happy holiday. I will refrain from a v3 release until you're back. No need to rush. And I'll be stepping back a bit too for a week or so.

**Comment by davidtheclark** [comment]
Done in #522.
